### ISSUE TITLE
Don't let a single slow peer kill performance of chain sync

### ISSUE THREAD

**Issue by gsalgado** [ethereum/py-evm#868]
### What is wrong?

When doing a chain sync, for each chain segment we split the missing block parts (bodies, receipts) in N batches (where N is the number of connected peers that have those block parts) and request each batch to a different peer. Quite often the peers will not reply with all requested items (because they implement a byte-size cap in responses), so we need to wait for replies from all peers in order to figure out what block parts we're still missing so that we can make another round of requests.

This obviously doesn't perform well if we have a slow peer.

### How can it be fixed

Need to keep track of the avg response time (or throughput?) for peers and request body parts only to the ones we deem fast enough (e.g. top 2/3rds ranked by response time, or below a certain response time threshold)

**Comment by carver** [comment]
> we need to wait for replies from **all peers** in order to figure out what block parts we're still missing

I'm still not understanding something about this. If we get the response from one peer, and it's partial, can't we immediately request the dropped pieces?

```
Request blocks 1-5 from Peer A
Request blocks 6-10 from Peer B
Peer A responds with blocks 1-2
<-- at this point, we can request blocks 3-5 from Peer A, before Peer B has responded, right?
```

**Comment by pipermerriam** [comment]
@carver I think yes, but I think that would be a separate optimization/issue.

**Comment by gsalgado** [comment]
@carver yeah, good point, we could do that!  It wouldn't solve the issue here, though, right?

**Comment by carver** [comment]
Sure, it doesn't strictly prevent us from hanging up on the particular blocks we asked the slow peer for, but it should speed up performance for all the other blocks.

Anyway, I'll create a new issue to track it.

**Comment by carver** [comment]
First pass on this is to track the throughputs and request a header count proportionally to each peer. So if peer A has throughput 1 block bodies per second and peer B does 3 block bodies per second, and we need 8 blocks, we would ask for 2 from peer A and 6 from peer B.

Second pass would probably be to add some kind of outlier cutoff mechanism, once we have collected enough throughput data.

**Comment by gsalgado** [comment]
@carver IIUC, unless we fix #879, the first pass you describe above may not be much of an improvement if we have a very slow peer as we'd still request blocks from it and have to wait for it on every iteration of the request-wait-reply-check-what's-missing loop

**Comment by carver** [comment]
> have to wait for it on every iteration of the request-wait-reply-check-what's-missing loop

Roughly true, until the speed difference is enough to reduce the headers we request from a peer is floored to 0.

If peer A is 4x slower than B, the cycle is something like:

- request 25 headers
- request 5 from A, request 20 from B
- receive 16 headers total
- request 0 from A, request 4 from B

If a peer is *really* slow, it will floor to 0 with even a decent number of headers remaining. Otherwise, the time it takes to get a response from a peer should balance out... Except for the way that fast peers return early with fewer headers than we asked for.

I may end up refactoring it significantly to get rid of the batching of requests altogether. If it takes too long, I may just punt the whole thing, because we shouldn't spend time optimizing something without data to show it will move the needle.

**Comment by carver** [comment]
Rough sketch for refactoring the fast sync process (which I don't think needs to happen in Grace Hopper):

### Header Sync
- roughly keep header sync the same (use our best peer exclusively, until we catch up to them), but...
- dump the headers into a priority queue (prioritized by low block number), and immediately request the next set
- if the queue gets too full (maybe `>= 2* MAX_REQUEST_HEADERS`), stop requesting headers
- if the queue gets empty enough (maybe `<= MAX_REQUEST_HEADERS`), start requesting headers again

### Block body sync
- keep track of which peers have a request in flight, and which are available
- loop through all available peers and make requests for any headers that have been downloaded but not processed
- when receiving block parts, re-add headers to the queue, for block parts that were not received
- can probably stop doing the speed-weighted requests to peers from #1009 

### Block persistence
- in block number order, persist a block once we have its header, body, and receipt, pausing if we are missing one of those pieces (and handling potential forks)

_Note to self: where are we persisting receipts?_

**Comment by carver** [comment]
I optimistically removed the milestone, does anyone think it should stay in Grace Hopper?

**Comment by pipermerriam** [comment]
I'm fine with removing from grace hopper.

**Comment by carver** [comment]
The problem of a slow peer hurting throughput on body/receipt requests has been solved, or will be once #1279 lands.

If we happen to pick a slow peer to do header sync, then that would still slow us down, but I think that should be a different issue. I merged it into #1218 since both issues require being able to switch header-syncing peers mid-stream.
